File: POTD/BFS/Cpp/3629. Minimum Jumps to Reach End via Prime Teleportation.cpp
class Solution {
private:
void sieve(vector<int>&isPrime,int maxi){
    isPrime[0]=isPrime[1]=false;

    for(int i=2;i*i<=maxi;i++){
        if(isPrime[i]){
            for(int j=i*i;j<=maxi;j+=i){
                isPrime[j]=false;
            }
        }
    }
}

vector<int> getPrimeFactors(int x,vector<int>&isPrime){
    vector<int> prime;

    for(int i=2;i*i<=x;i++){

        if(isPrime[i] && x%i==0){
            prime.push_back(i);

            while(x%i==0){
                x/=i;
            }
        }
    }


    if(x>1) prime.push_back(x);

    return prime;
}
public:
    int minJumps(vector<int>& nums) {
        int n=nums.size();

        int maxi=*max_element(begin(nums),end(nums));
        vector<int> isPrime(maxi+1,true);
        sieve(isPrime,maxi);


        unordered_map<int,vector<int>> teleMp;
        for(int i=0;i<n;i++){
            int x=nums[i];
            vector<int> primes=getPrimeFactors(x,isPrime);


            for(int p:primes){
                teleMp[p].push_back(i);
            }
        }


        queue<pair<int,int>> q;
        q.push({0,0});
        unordered_set<int> vis;
        vis.insert(0);


        while(!q.empty()){
            int idx=q.front().first;
            int steps=q.front().second;

            q.pop();

            if(idx==n-1) return steps;

            if(idx-1>=0 && !vis.count(idx-1)){
                q.push({idx-1,steps+1});
                vis.insert(idx-1);
            }

            if(idx+1<n && !vis.count(idx+1)){
                q.push({idx+1,steps+1});
                vis.insert(idx+1);
            }

            if(!isPrime[nums[idx]])continue;

            if(teleMp.count(nums[idx])){
                for(int pos:teleMp[nums[idx]]){
                    if(!vis.count(pos)){
                        q.push({pos,steps+1});
                        vis.insert(pos);
                    }
                }

                teleMp.erase(nums[idx]);
            }    
        }

        return -1;
    }
};

File: POTD/BFS/Python/3629. Minimum Jumps to Reach End via Prime Teleportation.py
class Solution:
    def sieve(self,nums:List[int],maxi:int):
        isPrime=[True]*(maxi+1)
        isPrime[0]=False
        isPrime[1]=False

        p=2
        while p*p<=maxi:
            if isPrime[p]:
                for mul in range(p*p,maxi+1,p):
                    isPrime[mul]=False
            
            p+=1
        
        return isPrime
    
    def getPrimeFactors(self,x:int,isPrime:List[bool]):
        primes=[]

        p=2
        while p*p<=x:
            if x%p==0 and isPrime[p]:
                primes.append(p)

                while x%p==0:
                    x//=p
            
            p+=1
        
        if x>1:
            primes.append(x)
        
        return primes

    def minJumps(self, nums: List[int]) -> int:
        n=len(nums)
        maxi=max(nums)

        isPrime=self.sieve(nums,maxi)
        mp=defaultdict(list)


        for i,x in enumerate(nums):
            primes=self.getPrimeFactors(x,isPrime)

            for p in primes:
                mp[p].append(i)
        
        dq : deque[tuple[int,int]]=deque()

        dq.append((0,0))
        vis=[False]*n
        vis[0]=True


        while dq:
            idx,steps=dq.popleft()

            if idx==n-1:
                return steps
            
            if idx-1>=0 and not vis[idx-1]:
                dq.append((idx-1,steps+1))
                vis[idx-1]=True

            if idx+1<n and not vis[idx+1]:
                dq.append((idx+1,steps+1))
                vis[idx+1]=True
            
            if not isPrime[nums[idx]]:
                continue
            
            if nums[idx] in mp:
                for pos in mp[nums[idx]]:
                    if not vis[pos]:
                        vis[pos]=True
                        dq.append((pos,steps+1))
                
                del mp[nums[idx]]
        
        return -1
